### PR TITLE
[BEAM-288] Improve UnboundedSource javadoc

### PR DESCRIPTION
**Pull request by mshields822**
R: @dhalperi

### CODE REVIEW

**Comment by mshields822** [comment]
ping

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 1]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
      * <p>For example, this could be sending acknowledgement requests to an external
      * data source such as Pub/Sub.
      *
-     * <p>This may be called from any thread, potentially at the same time as calls to the
-     * {@code UnboundedReader} that created it.
+     * <p>Note that:
+     * <ul>
+     * <li>This finalize method may be called from any thread, potentially at the same time as
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: "at the same time as" => "concurrently with"?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 2]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
...
+     * calls to the {@link UnboundedReader} it was created from.
+     * <li>Checkpoints will be finalized in the same order they are created from the
+     * {@link UnboundedReader}.
+     * <li>It is possible for multiple checkpoints created from the same {@link UnboundedReader}
+     * to be 'in-flight' (ie not yet finalized) simultaneously.
+     * <li>It is possible for a checkpoint to never be finalized.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: move this one up?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 3]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
      * <p>For example, this could be sending acknowledgement requests to an external
      * data source such as Pub/Sub.
      *
-     * <p>This may be called from any thread, potentially at the same time as calls to the
-     * {@code UnboundedReader} that created it.
+     * <p>Note that:
+     * <ul>
+     * <li>This finalize method may be called from any thread, potentially at the same time as
+     * calls to the {@link UnboundedReader} it was created from.
+     * <li>Checkpoints will be finalized in the same order they are created from the
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: "Checkpoints" => "Checkpoints that are finalized will"..

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 4]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
...
+     * {@link UnboundedReader}.
+     * <li>It is possible for multiple checkpoints created from the same {@link UnboundedReader}
+     * to be 'in-flight' (ie not yet finalized) simultaneously.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: and if there are, is there any guarantee that one will "win" or can they all be finalized?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 5]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
...
+     * <li>It is possible for a checkpoint to never be finalized.
+     * <li>If this call throws an exception then the entire checkpoint will be abandoned and the
+     * reader restarted from an earlier, successfully-finalized checkpoint.
+     * <li>The system ensures that if a checkpoint fails for any reason then no later checkpoint
+     * is allowed to be finalized without the reader first being restarted.
+     * <li>It is not safe to assume the {@link UnboundedReader} from which this checkpoint was
+     * created still exists at the time this method is called.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Probably, even stronger, that the checkpoint may be encoded and decoded so can have no pointers to in-memory state -- right?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 1 above)
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Spoke to @dpmills, going to update some of these comments.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 6]
```diff
@@ -110,8 +110,22 @@ public boolean requiresDeduping() {
...
+     * <li>This finalize method may be called from any thread, potentially at the same time as
+     * calls to the {@link UnboundedReader} it was created from.
+     * <li>Checkpoints will be finalized in the same order they are created from the
+     * {@link UnboundedReader}.
+     * <li>It is possible for multiple checkpoints created from the same {@link UnboundedReader}
+     * to be 'in-flight' (ie not yet finalized) simultaneously.
+     * <li>It is possible for a checkpoint to never be finalized.
+     * <li>If this call throws an exception then the entire checkpoint will be abandoned and the
+     * reader restarted from an earlier, successfully-finalized checkpoint.
+     * <li>The system ensures that if a checkpoint fails for any reason then no later checkpoint
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Drop "The system ensures"?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java [hunk 7]
```diff
@@ -205,10 +221,11 @@ public boolean requiresDeduping() {
...
      *
-     * <p>The elements read up until this is called will be processed together as a bundle. Once
-     * the result of this processing has been durably committed,
-     * {@link CheckpointMark#finalizeCheckpoint} will be called on the {@link CheckpointMark}
-     * object.
+     * <p>All elements read up until this method is called will be processed together as a bundle.
+     * (An element is considered 'read' if it could be returned by a call to {@link #getCurrent}.)
+     * Once the result of processing those elements and the returned checkpoint have been durably
+     * committed, {@link CheckpointMark#finalizeCheckpoint} will be called on the
+     * returned {@link CheckpointMark} object.
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: this sentence as written could be read to imply that finalizing will be called immediately after the durable commit. Consider rewording?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 1 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 3 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 2 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: I vote to leave it - I polished the text a bit.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 4 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 5 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: Add that in addition to the final comment, which is actually saying that the reader may have been swapped in between the checkpoint being taken and the finalize getting called even in normal operation.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 6 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java (hunk 7 above)
Comment (mshields822) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedSource.java]: added an eventually

**Comment by mshields822** [comment]
PTAL

**Comment by mshields822** [comment]
Not urgent, but ping so I can -d the branch.

**Pull request closed without merging**